### DISCUSSION TITLE
Issues deploying the project in django using nginx, gunicorn

### DISCUSSION THREAD

**Discussion by AdnaneSaber** [miguelgrinberg/python-socketio#883] [Q&A] [answered]
I would like to deploy this project which works perfectly on localhost or on python manage.py runserver command, when I try to deploy it I read on the documentation on the worker you should add these to gunicorn service which when I do -k eventlet -w 1, it returns 502 bad gateway This is what i'm getting based off https://github.com/miguelgrinberg/python-socketio/tree/main/examples/server/wsgi/django_example This is my gunicorn service
```
[Unit]
Description=gunicorn daemon
Requires=dev.socket
After=network.target
[Unit]
Description=gunicorn daemon
Requires=dev.socket
After=network.target

[Service]
User=myuser
Group=www-data
WorkingDirectory=/var/lib/jenkins/workspace/myproject
ExecStart=/var/lib/jenkins/workspace/myproject/env/bin/gunicorn \
          --access-logfile - \
          --workers 3 \
          -k eventlet \
          --bind unix:/run/dev.sock \
          myproject.wsgi:application
[Install]
WantedBy=multi-user.target
```
wsgi.py
```
import os

from django.core.wsgi import get_wsgi_application
import socketio

from app.views import sio

os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'myproject.settings')

wsgi_application = get_wsgi_application()
application = socketio.WSGIApp(sio, wsgi_application)
```
settings.py
```
...
WSGI_APPLICATION = 'myproject.wsgi.application'
...
```
If I remove the -k eventlet from the dev.service gunicorn service, the 502 bad gateway goes away and the page keeps loading forever and it returns 504 Gateway Time-out, and if I revert back wsgi.py to it's django by default boilerplate, it shows the page, but the websockets doesn't work and the page keeps sending https://example.com/socket.io/?EIO=4&transport=polling&t=opIU5ZA GET requests, that gets 502 bad gateway

**Comment by miguelgrinberg** [comment, marked as answer]
Please do not write bugs on this project when all you need is help making your application work. The discussion board is for this purpose.

**Reply by miguelgrinberg** [reply]
A 502 error is returned by your proxy server, not by this project or by Gunicorn. You need to look at your proxy to see why it is not able to forward requests to Gunicorn. The documentation shows how to run Gunicorn, by the way. If you are going to use eventlet, then you must have the `-k eventlet` option, removing it makes no sense, unless you already configure the application to not use eventlet.

**Reply by AdnaneSaber** [reply]
Actually the application us using eventlet on the views.py file like so:
```
...
sio = socketio.Server(async_mode='eventlet')
thread = None
...
```
And If I remove the argument from the gunicorn file, and revert back wsgi.py to the by default version, the website in general works which means the problem isn't from the nginx or gunicorn configuration, as I referred the problem is whether in the wsgi.py configuration or the gunicorn file, which I added into the discussion.

**Reply by miguelgrinberg** [reply]
You do not understand what I'm saying. If your application configures eventlet, then Gunicorn needs the eventlet worker. You can't say that it's fine because it works without it. It really doesn't, you can't just misconfigure eventlet.

You are also ignoring what I'm saying. There is no logic in this package that returns 502. If this is the error that you get (when the application and Gunicorn are correctly configured) then the error must be coming from your proxy server.

**Reply by AdnaneSaber** [reply]
I am not ignoring your words, and what I tried to say, is that i guess that it can't be coming from the proxy unless I forgot the add something or misadded something, because before trying to integrate python-socketio it was working fine, that's all i meant by when I remove the argument from gunicorn and revert wsgi.py it works, sorry if I less explained , this is my nginx configuration file located in **/etc/nginx/sites-available/myproject**
```

server {
    listen 80;
    server_name IP-V6;
    return 302 $scheme://example.com$request_uri;
}

server {
    server_name example.com;
    location = /favicon.ico { access_log off; log_not_found off; }
    location /static/ {
        root /var/lib/jenkins/workspace/myproject;
    }
    location / {
        include proxy_params;
        proxy_pass http://unix:/run/gunicorn.sock;
    }


    location /socket.io {
        include proxy_params;
        proxy_http_version 1.1;
        proxy_buffering off;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_pass http://localhost:8006/socket.io;
    }
}

```
and this is **runserver.py**
on **myproject/myapp/management/commands/runserver.py**
```
from django.core.management.commands.runserver import Command as RunCommand
import eventlet
from myapp.wsgi import application

class Command(RunCommand):
    help = 'Run the Socket.IO server'

    def handle(self, *args, **options):
        eventlet.wsgi.server(eventlet.listen(('', 8006)), application)

```

thank you for you time

**Reply by miguelgrinberg** [reply]
Are you using a unix socket or port 8006? Your nginx configuration has both, so either your Socket.IO or your non-Socket.IO endpoints are going to 502, I think.

**Reply by AdnaneSaber** [reply]
Before integrating socketio, i didn't use port 8006, but since on the docs I read that eventlet needs to listen on localhost with a port, so I added it, and probably it's the reason, can it be configured with unix socket ? don't mind me I am no expert in servers

**Reply by miguelgrinberg** [reply]
I'm not really sure I follow. Using the eventlet server is one thing, but you are using Gunicorn, right? So the eventlet web server doesn't really matter at all here. Also, what does the runserver command has to do with your production deployment if your service is configured to run Gunicorn?

Based on what I see, you should configure your proxy server to forward both HTTP and WebSocket to your unix socket, on which Gunicorn is listening. And you should use the eventlet worker with Gunicorn.

If you decide to drop eventlet, then change the async_mode to something else as well.

**Reply by AdnaneSaber** [reply]
Thank you so much @miguelgrinberg for your support.
I managed to make it work this way, i'll share the code
**_/etc/nginx/sites-available/myproject_**
```
server {
    server_name example.com;
    location = /favicon.ico { access_log off; log_not_found off; }
    location /static/ {
        root /var/lib/jenkins/workspace/myproject;
    }
    location / {
        include proxy_params;
        proxy_pass http://unix:/run/gunicorn.sock;
    }


    location /socket.io {
        include proxy_params;
        proxy_http_version 1.1;
        proxy_buffering off;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_pass http://unix:/run/gunicorn.sock;
    }
 }
```
**_/etc/systemd/system/gunicorn.service_**
```
[Unit]
Description=gunicorn daemon
Requires=gunicorn.socket
After=network.target

[Service]
User=myuser
Group=www-data
WorkingDirectory=/var/lib/jenkins/workspace/myproject
ExecStart=/var/lib/jenkins/workspace/myproject/env/bin/gunicorn \
          --log-file=- \
          -k eventlet \
          -w 1 \
          --bind unix:/run/gunicorn.sock \
          myproject.wsgi:application
[Install]
WantedBy=multi-user.target
```
myproject/wsgi.py
```
import os
os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'myproject.settings')
from django.core.wsgi import get_wsgi_application
import socketio
from websocketapp.views import sio
wsgi_application = get_wsgi_application()
application = socketio.WSGIApp(sio, wsgi_application)
```
i think the environment variable set should be before the socketio imports for some reason.
This is the config that worked for me, thank you so much again and I loved the **python-socketio** project, i hope this discussion could help someone in the future.
